pkg/deploy/identity-provider/deployment_keycloak.go

```diff
@@ -485,8 +489,37 @@ func getSpecKeycloakDeployment(
 		keycloakEnv = append(keycloakEnv, envvar)
 	}
 
-	command := addCertToTrustStoreCommand + addProxyCliCommand + applyProxyCliCommand + " && " + changeConfigCommand +
-		" && /opt/jboss/docker-entrypoint.sh -b 0.0.0.0 -c standalone.xml"
+	var enableFixedHostNameProvider string
+	if deployContext.CheCluster.Spec.Server.UseServiceHostNames {
+		if cheFlavor == "che" {
+			keycloakURL, err := url.Parse(deployContext.CheCluster.Status.KeycloakURL)
+			if err != nil {
+				return nil, err
+			}
+			hostname := keycloakURL.Hostname()
+			enableFixedHostNameProvider = " && echo 'Use fixed hostname provider to make working internal network requests' && " +
+			"echo -e \"embed-server --server-config=standalone.xml --std-out=echo \n" +
+			"/subsystem=keycloak-server/spi=hostname:write-attribute(name=default-provider, value=\"fixed\") \n" +
+			"/subsystem=keycloak-server/spi=hostname/provider=fixed:write-attribute(name=properties.hostname,value=\"" + hostname + "\") \n"
+			if deployContext.CheCluster.Spec.Server.TlsSupport {
+				enableFixedHostNameProvider += "/subsystem=keycloak-server/spi=hostname/provider=fixed:write-attribute(name=properties.httpsPort,value=\"443\") \n"  +
+				"/subsystem=keycloak-server/spi=hostname/provider=fixed:write-attribute(name=properties.alwaysHttps,value=\"true\") \n"
+			} else {
+				enableFixedHostNameProvider += "/subsystem=keycloak-server/spi=hostname/provider=fixed:write-attribute(name=properties.httpPort,value=\"80\") \n"
+			}
+			enableFixedHostNameProvider += "stop-embedded-server\" > " + jbossDir  + "/use_fixed_hostname_provider.cli && " +
+			jbossCli + " --file=" + jbossDir + "/use_fixed_hostname_provider.cli "
+		}
+		if cheFlavor == "codeready" {
+			keycloakEnv = append(keycloakEnv, corev1.EnvVar{
+				Name: "KEYCLOAK_FRONTEND_URL",
+				Value: deployContext.CheCluster.Status.KeycloakURL + "/auth",
```

Comment:
Isn't it the Keycloak *internal* URL that we want here ?

Reply:
Keycloak includes `issuer` url inside each token like one of the fields. For our case user dashboard, using keycloak adapters, retrieves tokens from keycloak. And user dashboard uses public host url for this purpose, because user dashboard is frontend and can't use internal network. After that we have decoded 'issuer' field with `public hostname` url inside token. Then che-server needs to reuse this token from user dashboard to make some server side requests(retrieve user profile, link user account, get openshift 4 token, delete user from keycloak and so on...). che-server should use internal url to make requests, but requests with token generated for `public hostname url` fails due keycloak validation. Keycloak compare request url(with internal network value) to token `issuer` and fail che-server request. That's why we set up to keycloak `KEYCLOAK_FRONTEND_URL` env. Keycloak in this mode compare token issuer to the `KEYCLOAK_FRONTEND_URL` and che-server request works fine.